src/compiler/core.ts

```diff
@@ -1489,45 +1528,264 @@ namespace ts {
         return headChain;
     }
 
-    export function compareValues<T>(a: T, b: T): Comparison {
-        if (a === b) return Comparison.EqualTo;
-        if (a === undefined) return Comparison.LessThan;
-        if (b === undefined) return Comparison.GreaterThan;
-        return a < b ? Comparison.LessThan : Comparison.GreaterThan;
+    export function equateValues<T>(a: T, b: T) {
+        return a === b;
+    }
+
+    /**
+     * Compare the equality of two strings using a case-sensitive ordinal comparison.
+     *
+     * Case-sensitive comparisons compare both strings one code-point at a time using the integer
+     * value of each code-point after applying `toUpperCase` to each string. We always map both
+     * strings to their upper-case form as some unicode characters do not properly round-trip to
+     * lowercase (such as `ẞ` (German sharp capital s)).
+     */
+    export function equateStringsCaseInsensitive(a: string, b: string) {
+        return a === b
+            || a !== undefined
+            && b !== undefined
+            && a.toUpperCase() === b.toUpperCase();
+    }
+
+    /**
+     * Compare the equality of two strings using a case-sensitive ordinal comparison.
+     *
+     * Case-sensitive comparisons compare both strings one code-point at a time using the
+     * integer value of each code-point.
+     */
+    export function equateStringsCaseSensitive(a: string, b: string) {
+        return equateValues(a, b);
     }
 
-    export function compareStrings(a: string, b: string, ignoreCase?: boolean): Comparison {
+    function compareComparableValues(a: string, b: string): Comparison;
+    function compareComparableValues(a: number, b: number): Comparison;
+    function compareComparableValues(a: string | number, b: string | number) {
+        return a === b ? Comparison.EqualTo :
+            a === undefined ? Comparison.LessThan :
+            b === undefined ? Comparison.GreaterThan :
+            a < b ? Comparison.LessThan :
+            Comparison.GreaterThan;
+    }
+
+    /**
+     * Compare two values for their order relative to each other.
+     */
+    export function compareValues(a: number, b: number) {

```

Comment:
Rename to `compareNumbers`?

Reply:
I'd rather leave the name as is for the time being. Too many engineers on the team are used to it having this name. I've added an additional comment indicating where to find the string comparison functions, to ease the transition.